qiskit/circuit/instruction.py

```diff
@@ -379,3 +380,30 @@ def repeat(self, n):
 
         instruction.definition = [(self, qargs[:], cargs[:])] * n
         return instruction
+
+
+def deprecate_arguments(kwarg_map):
+    """Decorator to automatically alias deprecated agrument names and warn upon use."""
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            if kwargs:
+                _rename_kwargs(func.__name__, kwargs, kwarg_map)
+            return func(*args, **kwargs)
+        return wrapper
+    return decorator
+
+
+def _rename_kwargs(func_name, kwargs, kwarg_map):
+    for old_arg, new_arg in kwarg_map.items():
+        if old_arg in kwargs:
+            if new_arg in kwargs:
+                raise TypeError('{} received both {} and {} (deprecated).'.format(
+                    func_name, new_arg, old_arg))
+
+            warnings.warn('{} keyword argument {} is deprecated and '
+                          'replaced with {}.'.format(
+                              func_name, old_arg, new_arg),
+                          DeprecationWarning)
```

Comment:
We probably should change this to have `stacklevel=3` so it shows the function call instead of this line (or the `_rename_kwargs` call in the decorator.

Reply:
Added `stacklevel=3` in 5c35f4cdd841db632097cb43bb13486184b58131.